cupyx/scipy/signal/_iir_utils.py

```diff
@@ -0,0 +1,252 @@
+        for(int i = 1; i <= min(k, level); i++) {
+            float k_value = group_start[low_bound - i];
+            const float* k_factors = factors + (m + k) * (i - 1) + k;
+            float factor = k_factors[rel_pos];
+            carry += k_value * factor;
+        }
+
+        group_start[pos] += carry;
+        __syncthreads();
+    }
+
+    if(pos >= m - k) {
+        if(carries != NULL) {
+            group_carries[pos - (m - k)] = group_start[pos];
+        }
+    }
+
+}
+
+__global__ void correct_carries(
+    const int m, const int k, const int n_blocks, const int carries_stride,
+    const int offset, const float* factors, float* carries) {
+
+    int idx = threadIdx.x;
+    int pos = idx + (m - k);
+    float* row_carries = carries + carries_stride * blockIdx.x;
+
+    for(int i = offset; i < n_blocks; i++) {
+        float* this_carries = row_carries + k * (i + (1 - offset));
+        float* prev_carries = row_carries + k * (i - offset);
+
+        float carry = 0.0;
+        for(int j = 1; j <= k; j++) {
+            const float* k_factors = factors + (m + k) * (j - 1) + k;
+            float factor = k_factors[pos];
+            float k_value = prev_carries[k - j];
+            carry += factor * k_value;
+        }
+
+        this_carries[idx] += carry;
+        __syncthreads();
+    }
+}
+
+__global__ void second_pass_iir(
+        const int m, const int k, const int n, const int carries_stride,
+        const int n_blocks, const int offset, const float* factors,
+        float* carries, float* out) {
+
+    int idx = blockDim.x * (blockIdx.x % n_blocks) + threadIdx.x;
+    idx += offset * m;
+
+    int row_num = blockIdx.x / n_blocks;
+    int n_group = idx / m;
+    int pos = idx % m;
+
+    if(idx >= n) {
+        return;
+    }
+
+    float* out_off = out + row_num * n;
+    float* carries_off = carries + row_num * carries_stride;
+    const float* prev_carries = carries_off + (n_group - offset) * k;
+
+    float carry = 0.0;
+    for(int i = 1; i <= k; i++) {
+        const float* k_factors = factors + (m + k) * (i - 1) + k;
+        float factor = k_factors[pos];
+        float k_value = prev_carries[k - i];
+        carry += factor * k_value;
+    }
+
+    out_off[idx] += carry;
+}
+
+"""
+
+IIR_MODULE = cupy.RawModule(
+    code=IIR_KERNEL, options=('-std=c++11',),
+    name_expressions=['compute_correction_factors',
+                      'correct_carries',
+                      'first_pass_iir',
+                      'second_pass_iir'])
+
+
+def collapse_2d(x, axis):
+    x = cupy.moveaxis(x, axis, -1)
+    x_shape = x.shape
+    x = x.reshape(-1, x.shape[-1])
+    if not x.flags.c_contiguous:
+        x = x.copy()
+    return x, x_shape
+
+
+def apply_iir(x, a, axis=-1, zi=None):
+    # GPU throughput is faster when using single precision floating point
+    # numbers
+
+    x = x.astype(cupy.float32)
+    a = a.astype(cupy.float32)
+
+    if zi is not None:
+        zi = zi.astype(cupy.float32)
```

Comment:
@takagi, should we always resort to float32 for speed purposes, or should we specify in `lfilter` docstring that works best with `float32` and add type templates to the kernels?

Reply:
Please note it on the document and leave the dtype handling as the same with SciPy.